src/main/solution_space/path_finder.py

```diff
@@ -0,0 +1,166 @@
+        user_diffs_to_goal = get_diffs_number(user_code.file_with_code, goal.code.file_with_code)
+        candidates = []
+        vertices = self._graph.get_traversal()
+        vertices.remove(self._graph.start_vertex)
+        for vertex in vertices:
+            # We don't want to add to result the same vertex
+            if are_asts_equal(user_code.ast, vertex.code.ast):
+                continue
+            diffs = get_diffs_number(vertex.code.file_with_code, goal.code.file_with_code)
+            if diffs <= user_diffs_to_goal:
+                candidates.append(vertex)
+        return self.__choose_best_vertex(user_code, user, vertices)
+
+    # Choose the best way to go to the goal
+    # For example, if we have a good way through the graph, we should advise it,
+    # but if don't we would advise going directly to the goal
+    @staticmethod
+    def __go_through_graph(user_code: Code, graph_vertex: Vertex, goal: Vertex,
+                           median_of_diff_number_to_goal: int) -> bool:
+        # Todo: use graph_vertex too
+        diffs_to_goal = get_diffs_number(user_code.file_with_code, goal.code.file_with_code)
+        border_go_through_graph = median_of_diff_number_to_goal * (1 - DIFFS_PERCENT_GO_THROUGH_GRAPH)
+        if diffs_to_goal <= border_go_through_graph:
+            return False
+        return True
+
+    # Call gumTreeDiff methods
+    @staticmethod
+    def __apply_minimal_actions_number(user_code: Code, dst_code: Code) -> Code:
+        str_code = apply_diffs(user_code.file_with_code, dst_code.file_with_code)
+        ast = get_canonicalized_form(str_code)
+        # Todo: get rate
+        return Code(ast)
+
+
+# Todo: rename it???
+class Candidate:
+    def __init__(self, user_code: Code, vertex: Vertex, user: User, distance: Optional[int] = None):
+        self._vertex = vertex
+        self._distance = distance if distance \
+            else get_diffs_number(user_code.file_with_code, vertex.code.file_with_code)
+        # Todo: get actual vertex profile
+        self._profile = self.__init_profile(user)
+        self._users_count = len(vertex.get_unique_users())
+
+    @property
+    def vertex(self) -> Vertex:
+        return self._vertex
+
+    @property
+    def distance(self) -> int:
+        return self._distance
+
+    @property
+    def profile(self) -> Profile:
+        return self._profile
+
+    @property
+    def users_count(self) -> int:
+        return self._users_count
+
+    def __init_profile(self, user: User) -> Profile:
+        # Todo: add user handling
+        ages, experiences = [], []
+        SolutionGraph.update_ages_and_experiences(self._vertex.code_info_list, ages, experiences)
+        return Profile(calculate_safety_median(ages, default_value=DEFAULT_VALUES.AGE.value),
+                       calculate_median_for_objects(experiences, default_value=DEFAULT_VALUES.EXPERIENCE))
```

Comment:
Why don't you call calculate_median_of_profile_info(self)? However, i think we shouldn't init Ptofil with median right now, because we want to have as simple as possible version of PathFinder. Let's init it with None and leave medians for other pr?